### PR TITLE
Add more information to authgear errors

### PR DESCRIPTION
**Pull request by tung2744**

### CODE REVIEW

File: Sources/AuthgearError.swift [hunk 1]
```diff
@@ -24,12 +24,88 @@ public enum AuthgearError: Error {
...
+        case let .biometricNoPasscode(err):
+            info["error"] = err.localizedDescription
+        case let .biometricNoEnrollment(err):
+            info["error"] = err.localizedDescription
+        case let .biometricLockout(err):
+            info["error"] = err.localizedDescription
+        case let .invalidJWT(jwt):
+            info["jwt"] = jwt
+        case let .error(err):
+            info["error"] = err.localizedDescription
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:84]: For cases that wrap an `Error`, I think we should cast the `Error` to `NSError`, and then construct the error info as follows:

```
{
  "domain": NSError.domain,
  "code": NSError.code,
  "info": NSError.info,
  "localizedDescription": NSError.localizedDescription
}
```

That also means we should make all our error structs/classes (including ServerError, OAuthError, and LatteError) implements CustomNSError too.

I would like to know now that AuthgearError implements CustomNSError, if we cast it to NSError, would the domain, code, and info correct?

Comment (tung2744) [reply Sources/AuthgearError.swift:84]: @louischan-oursky Yes. Actually, all `Error` have a default implementation of `NSError`, which use its class name domain, and the enumerator as code. However we cannot control the code and have no idea how to map it back to the exact enum. Implementing CustomNSError let us customize this behavior.

File: Sources/LatteError.swift [hunk 2]
```diff
@@ -1,5 +1,6 @@
 import Foundation
 
+// Implement LocalizedError to customize localizedDescription
 enum LatteError: LocalizedError {
```
[the remark this answers is not in the document]
Comment (tung2744) [reply Sources/LatteError.swift:4]: Maybe we implement both so that we have customized error message with CustomNSError

**Comment by tung2744** [comment]
@louischan-oursky Updated thx!

File: Sources/AuthgearError.swift [hunk 3]
```diff
@@ -71,22 +71,33 @@ public enum AuthgearError: CustomNSError {
...
-            info["error"] = err.localizedDescription
+            info["error"] = makeErrorInfo(err)
         case let .biometricLockout(err):
-            info["error"] = err.localizedDescription
+            info["error"] = makeErrorInfo(err)
         case let .invalidJWT(jwt):
             info["jwt"] = jwt
         case let .error(err):
-            info["error"] = err.localizedDescription
+            info["error"] = makeErrorInfo(err)
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:84]: Do not nest.

File: Sources/AuthgearError.swift [hunk 4]
```diff
@@ -3,7 +3,7 @@ import Foundation
 import LocalAuthentication
 import SafariServices
 
-public enum AuthgearError: Error {
+public enum AuthgearError: CustomNSError {
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:6]: For all our error data type, implements LocalizedError and CustomNSError.

File: Sources/AuthgearError.swift [hunk 5]
```diff
@@ -24,12 +24,88 @@ public enum AuthgearError: Error {
...
+        case let .error(err):
+            info["error"] = err.localizedDescription
+        default:
+            break
+        }
+        return info
+    }
 }
 
 public struct OAuthError: Error, Decodable {
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:103]: errorDecription: `${error}: ${errorDescription}`

File: Sources/AuthgearError.swift [hunk 6]
```diff
@@ -24,12 +24,88 @@ public enum AuthgearError: Error {
...
+            userInfo["errorDescription"] = errorDescription
+        }
+        if let errorUri = self.errorUri {
+            userInfo["errorUri"] = errorUri
+        }
+        return userInfo
+    }
 }
 
 public struct ServerError: Error, Decodable {
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:122]: errorDescription: `${reason}: ${message}`

File: Sources/AuthgearError.swift [hunk 7]
```diff
@@ -24,12 +24,88 @@ public enum AuthgearError: Error {
     case unauthenticatedUser
     case publicKeyNotFound
     case error(Error)
+
```
Comment (louischan-oursky) [inline Sources/AuthgearError.swift:27]: errorDescription: `${variantName}`

File: Sources/AuthgearError.swift (hunk 5 above)
Comment (tung2744) [reply Sources/AuthgearError.swift:103]: I found that we have property conflict with `LocalizedError` in OAuthError.
I think for OAuthError I will implement `localizedDescription` of `Error` instead.

**Comment by tung2744** [comment]
@louischan-oursky Updated thx!

**Pull request merged**